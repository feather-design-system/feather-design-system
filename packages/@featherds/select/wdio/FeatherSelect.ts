const OPTION = "[data-ref-id='feather-select-list'] li";
export class FeatherSelect {
  private selectSelector: string;
  get select() {
    return $(this.selectSelector);
  }
  constructor(_select: string) {
    this.selectSelector = _select;
  }

  async selectByText(text: string) {
    const runInBrowser = function (argument: WebdriverIO.Element) {
      argument.click();
    };
    const select = await this.select;
    await browser.execute(runInBrowser, select);

    await $(OPTION).waitForDisplayed({ timeout: 60000 });

    const items = await $$(OPTION);
    const textArray = await Promise.all(items.map((item) => item.getText()));
    const itemIndex = textArray.indexOf(text);
    if (itemIndex > -1) {
      const result = await items[itemIndex].getText();
      await items[itemIndex].click();
      return result.trim();
    }

    throw new Error(
      "Unable to select item '" +
        text +
        "' from " +
        this.selectSelector +
        " list"
    );
  }

  async selectByIndex(index: number) {
    const runInBrowser = function (argument: WebdriverIO.Element) {
      argument.click();
    };
    const select = await this.select;
    await browser.execute(runInBrowser, select);
    await $(OPTION).waitForDisplayed({ timeout: 60000 });
    const item = await $$(OPTION)[index];
    if (item) {
      const result = await item.getText();
      await item.click();
      return result.trim();
    }
    throw new Error(
      "Unable to select index " +
        index +
        " from " +
        this.selectSelector +
        " list"
    );
  }
  async getValue() {
    return await this.select.getValue();
  }
}
