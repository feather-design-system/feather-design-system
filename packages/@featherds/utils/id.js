const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
