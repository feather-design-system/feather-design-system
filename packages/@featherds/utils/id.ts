const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
