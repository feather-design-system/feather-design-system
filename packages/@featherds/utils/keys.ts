const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
