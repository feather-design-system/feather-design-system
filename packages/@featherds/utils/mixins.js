var translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (key) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};
export { translateKey };
