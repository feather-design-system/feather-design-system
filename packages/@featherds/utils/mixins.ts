const translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (this: IComponent, key: string) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};

interface IComponent {
  translationKey: string;
  $t: (s: string) => void;
}
export { translateKey };
