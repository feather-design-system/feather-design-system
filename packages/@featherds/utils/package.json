{
  "name": "@featherds/utils",
  "version": "0.12.31",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {
    "compile": "tsc --declaration"
  },
  "keywords": [
    "vue",
    "FeatherUtils"
  ]
}
