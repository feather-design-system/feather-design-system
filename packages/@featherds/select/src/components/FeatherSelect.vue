<template>
  <div v-bind="inherittedAttrs" class="feather-select-container">
    <FeatherMenu
      no-expand
      :open="showMenu"
      @outside-click="handleOutsideClick"
      @trigger-click="handleTriggerClick"
      @close="closeMenu"
      class="feather-select-menu-container"
      data-ref-id="feather-select-menu-container"
    >
      <template v-slot:trigger>
        <InputWrapper
          :for="inputId"
          :raised="raised"
          :focused="hasFocus"
          :show-clear="showClear"
          @clear="handleClear"
          menu-trigger
          class="feather-select-wrapper"
          :class="{ focused: hasFocus }"
        >
          <template v-slot:pre>
            <slot name="pre" />
          </template>
          <input
            v-bind="inputAttrs"
            class="feather-select-input"
            data-ref-id="feather-select-input"
            v-on="inputListeners"
            ref="input"
          />
          <template v-slot:post>
            <FeatherIcon
              class="feather-select-icon"
              :class="{ rotate: showMenu }"
              :icon="icon"
            ></FeatherIcon>
          </template>
        </InputWrapper>
      </template>
      <List
        ref="list"
        data-ref-id="feather-select-list"
        :label="label"
        :options="options"
        :text-prop="textProp"
        :active-index="activeIndex"
        @select="handleSelect"
        @keydown="handleKeyDown"
      >
      </List>
    </FeatherMenu>
    <InputSubText :id="subTextId"></InputSubText>
  </div>
</template>
<script lang="ts">
import {
  InputWrapper,
  InputSubText,
  InputWrapperProps,
  InputSubTextProps,
  useInputWrapper,
  useInputSubText,
  useInputInheritAttrs,
} from "@featherds/input-helper";
import { FeatherIcon } from "@featherds/icon";
import KeyboardArrowDown from "@featherds/icon/navigation/ExpandMore";
import { FeatherMenu } from "@featherds/menu";
import { getSafeId } from "@featherds/utils/id";
import { KEYCODES } from "@featherds/utils/keys";
import List from "./List.vue";
import { useValidation } from "@featherds/input-helper";
import { _setTimeout } from "@featherds/utils/setTimeout";
import { computed, defineComponent, PropType, Ref, toRef } from "vue";
import { ISelectItemType } from "./types";
export const props = {
  ...InputWrapperProps,
  ...InputSubTextProps,
  modelValue: {
    type: Object as PropType<ISelectItemType>,
    required: false,
  },
  textProp: {
    type: String,
    default: "_text",
  },
  options: {
    type: Array as PropType<ISelectItemType[]>,
    default: () => {
      return [] as ISelectItemType[];
    },
  },
  schema: {
    type: Object,
    required: false,
  },
} as const;
export const emits = {
  "update:modelValue": (v: ISelectItemType | undefined) => true,
};
export default defineComponent({
  model: {
    prop: "modelValue",
    event: "update:modelValue",
  },
  emits,
  props,
  setup(props, context) {
    useInputSubText(props);
    useInputWrapper(props);
    const inputId = computed(() => {
      return getSafeId("feather-select-input");
    });

    const { validate } = useValidation(
      inputId,
      toRef(props, "modelValue"),
      props.label,
      props.schema as Record<string, unknown>,
      toRef(props, "error") as Ref<string>
    );

    return {
      inputId,
      validate,
      ...useInputInheritAttrs(context.attrs as Record<string, unknown>),
    };
  },
  data() {
    return {
      hasFocus: false,
      showMenu: false,
      charsSoFar: "",
      internalValue: this.modelValue,
      delayTimeout: 0,
    };
  },
  computed: {
    showClear() {
      return !!this.modelValue;
    },
    subTextId() {
      return getSafeId("feather-select-description");
    },
    inputAttrs() {
      return {
        id: this.inputId,
        "aria-haspopup": "listbox" as
          | "grid"
          | "dialog"
          | "menu"
          | "listbox"
          | "tree"
          | undefined
          | boolean
          | "true"
          | "false",
        "aria-invalid":
          (this.$attrs["aria-invalid"] as boolean) || !!this.error,
        value: this.valueText,
        readonly: true,
        disabled: this.disabled,
        "aria-disabled": this.disabled,
        "aria-describedby": ((this.$attrs["aria-describedby"] as string) || "")
          .split(" ")
          .concat([this.subTextId])
          .filter(Boolean)
          .join(" "),
      };
    },
    inputListeners() {
      return {
        focus: this.handleInputFocus,
        blur: this.handleInputBlur,
        keydown: this.handleKeyDown,
      };
    },
    raised() {
      return !!this.internalValue || this.hasFocus;
    },
    valueText() {
      if (this.internalValue && this.internalValue[this.textProp]) {
        return this.internalValue[this.textProp] as string;
      }
      return "";
    },
    activeIndex() {
      if (this.internalValue && this.internalValue[this.textProp]) {
        const iv = this.internalValue as ISelectItemType;
        const found = this.options.filter(
          (x) => x[this.textProp] === iv[this.textProp]
        );
        if (found && found.length) {
          return this.options.indexOf(found[0]);
        }
      }
      return -1;
    },
    icon: () => KeyboardArrowDown,
  },
  watch: {
    showMenu(v) {
      if (v) {
        if (!this.internalValue) {
          this.select(this.options[0]);
        }
        this.$nextTick(() => {
          (this.$refs.input as HTMLInputElement).focus();
        });
      } else {
        this.emitSelection();
      }
    },
    modelValue(v) {
      this.internalValue = v;
    },
  },
  methods: {
    closeMenu() {
      this.showMenu = false;
      this.$nextTick(() => {
        (this.$refs.input as HTMLInputElement).focus();
      });
    },
    handleClear() {
      this.handleSelect(undefined);
      this.emitSelection();
    },
    handleInputFocus() {
      if (this.disabled) {
        return;
      }
      if (!this.hasFocus) {
        this.hasFocus = true;
      }
    },
    handleInputBlur() {
      if (this.hasFocus && !this.showMenu) {
        this.hasFocus = false;
        this.validate();
      }
    },
    handleTriggerClick() {
      if (this.disabled) {
        return;
      }
      this.showMenu = true;
      this.hasFocus = true;
    },
    handleOutsideClick() {
      this.showMenu = false;
      this.hasFocus = false;
    },
    handleSelect(option: ISelectItemType | undefined) {
      this.select(option);
      this.showMenu = false;
      (this.$refs.input as HTMLInputElement).focus();
    },
    select(option: ISelectItemType | undefined) {
      this.internalValue = option;
    },
    emitSelection() {
      this.$emit("update:modelValue", this.internalValue);
    },
    handleKeyDown(e: KeyboardEvent) {
      //enter
      if (e.keyCode === KEYCODES.ENTER) {
        e.preventDefault();
        this.showMenu = !this.showMenu;
        //if hidden focus button
        if (!this.showMenu) {
          this.$nextTick(() => {
            (this.$refs.input as HTMLInputElement).focus();
          });
        }
      }
      //esc
      else if (e.keyCode === KEYCODES.ESCAPE) {
        this.closeMenu();
        e.stopPropagation();
      } else if (e.keyCode === KEYCODES.DOWN) {
        //down
        e.preventDefault();

        if (this.activeIndex + 1 < this.options.length) {
          this.select(this.options[this.activeIndex + 1]);
        }
        this.showMenu = true;
      } else if (e.keyCode === KEYCODES.UP) {
        //up
        e.preventDefault();

        if (this.activeIndex - 1 >= 0) {
          this.select(this.options[this.activeIndex - 1]);
        }
        this.showMenu = true;
      } else if (e.keyCode === KEYCODES.HOME) {
        //home
        e.preventDefault();
        this.select(this.options[0]);
        this.showMenu = true;
      } else if (e.keyCode === KEYCODES.END) {
        //home
        e.preventDefault();
        this.select(this.options[this.options.length - 1]);
        this.showMenu = true;
      } else if (this.showMenu) {
        const char = String.fromCharCode(e.keyCode);
        this.charsSoFar += char;
        this.searchAfterDelay();
      }
    },
    searchAfterDelay() {
      clearTimeout(this.delayTimeout);
      this.delayTimeout = _setTimeout(() => {
        const found = this.options.filter(
          (x) =>
            x[this.textProp] &&
            (x[this.textProp] as string)
              .toLowerCase()
              .indexOf(this.charsSoFar.toLowerCase()) === 0
        );
        if (found && found.length) {
          this.select(found[0]);
        }
        this.charsSoFar = "";
      }, 250);
    },
  },
  components: {
    InputWrapper,
    InputSubText,
    FeatherMenu,
    List,
    FeatherIcon,
  },
});
</script>

<style lang="scss" scoped>
@import "@featherds/styles/mixins/typography";
@import "@featherds/styles/themes/variables";

.feather-select-input {
  border: none;
  padding: 0;
  padding-left: 0.25rem;
  margin: 0;
  display: inline-block;
  vertical-align: middle;
  height: 100%;
  background-color: transparent;
  @include body-small();
  line-height: 1.5em; //IE has an issue with the mixin lineheight when typing
  color: var($secondary-text-on-surface);
  flex: 1;
  pointer-events: none;
  cursor: pointer;
  min-width: 0;
  &:focus,
  &:active {
    border-color: transparent;
    outline: 0;
  }
  &:hover {
    cursor: pointer;
  }
}
.feather-select-icon {
  flex: none;
  font-size: 1.25rem;
  height: 1.25rem;
  width: 1.25rem;
  align-self: center;
  color: var($secondary-text-on-surface);
  transition: transform 280ms ease-in-out;
  transform-origin: center center;
  &.rotate {
    transform: rotate(180deg);
  }
}
.feather-select-menu-container {
  width: 100%;
  position: relative;
  :deep(.feather-menu-dropdown) {
    min-width: 100%;
  }
}
.feather-select-wrapper {
  &.focused .feather-select-input {
    color: var($primary-text-on-surface);
  }
  &.disabled {
    .feather-select-input {
      color: var($disabled-text-on-surface);
    }
    .feather-select-icon {
      color: var($disabled-text-on-surface);
    }
  }
  &:hover {
    cursor: pointer;
    :deep(.feather-input-wrapper) {
      cursor: pointer !important;
    }
  }
  :deep(.feather-input-wrapper) {
    align-items: center;
  }
}
</style>
