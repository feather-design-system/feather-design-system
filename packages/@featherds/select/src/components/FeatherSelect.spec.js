import FeatherSelect from "./FeatherSelect.vue";
import { shallowMount, mount } from "@vue/test-utils";
import * as id from "@featherds/utils/id";
import axe from "@featherds/utils/test/axe";
import "@featherds/input-helper/test/MutationObserver";
import { nextTick } from "vue";

jest.spyOn(id, "getSafeId").mockImplementation((x) => x);

const getProps = (props) => {
  return {
    label: "Users",
    ...props,
  };
};

const options = [
  {
    _text: "Item 1",
  },
  {
    _text: "Item 2",
  },
];
const getWrapper = function (options = {}) {
  options.props = getProps(options.props || {});
  options.attachTo = document.body;
  return shallowMount(FeatherSelect, options);
};
const getFullWrapper = function (options = {}) {
  document.body.innerHTML = "";
  options.props = getProps(options.props || {});
  options.attachTo = document.body;
  return mount(FeatherSelect, options);
};

describe("FeatherSelect.vue", () => {
  it("should raise the label when it has a value", () => {
    const modelValue = {
      _text: "Test1",
    };
    const wrapper = getFullWrapper({
      props: {
        modelValue,
      },
    });
    expect(wrapper.wrapperElement).toMatchSnapshot();
  });
  it("should not raise the label when its empty", () => {
    const wrapper = getFullWrapper();
    expect(wrapper.wrapperElement).toMatchSnapshot();
  });
  it("should render in error state", () => {
    const wrapper = getFullWrapper({
      props: {
        error: "ERROR",
      },
    });
    expect(wrapper.wrapperElement).toMatchSnapshot();
  });
  it("should render in disabled state", () => {
    const wrapper = getFullWrapper({
      props: {
        disabled: true,
      },
    });
    expect(wrapper.wrapperElement).toMatchSnapshot();
  });
  it("should open the menu on enter", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.enter");
    expect(wrapper.vm.showMenu).toBe(true);
  });
  it("should close the menu on esc and emit update:modelValue", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.enter");
    await wrapper.find("input").trigger("keydown.esc");
    expect(wrapper.vm.showMenu).toBe(false);
    expect(wrapper.emitted("update:modelValue")[0][0]).toStrictEqual(
      options[0]
    );
  });

  it("should clear the value when clear icon is clicked", async () => {
    const modelValue = {
      _text: "Test1",
    };
    const wrapper = getFullWrapper({
      props: {
        modelValue,
        clear: "Clear Value",
      },
    });
    await wrapper
      .find("[data-ref-id='feather-form-element-clear']")
      .trigger("click");
    expect(wrapper.emitted("update:modelValue")[0][0]).toBeUndefined();
  });
  it("should open and select first item on first down", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.down");
    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[0]);
  });
  it("should open and select first item on first up", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.up");
    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[0]);
  });
  it("should move selection down when down arrow is pressed and menu is open", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[0],
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.down");
    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[1]);
  });
  it("should not emit update:model till menu is closed", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[0],
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.down");
    expect(wrapper.vm.showMenu).toBe(true);
    await wrapper.find("input").trigger("keydown.esc");
    expect(wrapper.vm.showMenu).toBe(false);
    expect(wrapper.emitted("update:modelValue")[0][0]).toStrictEqual(
      options[1]
    );
  });
  it("should not wrap to the top when down is pressed on the last option", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[options.length - 1],
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.down");
    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[options.length - 1]);
  });
  it("should move selection up when up arrow is pressed and menu is open", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[options.length - 1],
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.up");
    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[options.length - 2]);
  });
  it("should not wrap to the bottom when up is pressed on the first option", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[0],
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.up");
    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[0]);
  });
  it("should move selection to first item  when home is pressed and menu is open", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[options.length - 1],
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.home");
    await nextTick();
    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[0]);
  });
  it("should move selection to last item when end is pressed and menu is open", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
        modelValue: options[0],
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.end");
    expect(wrapper.vm.showMenu).toBe(true);
    expect(wrapper.vm.internalValue).toStrictEqual(options[options.length - 1]);
  });
  it("should remain focused when item is selected via click and emit update:modelValue", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.enter"); //open menu
    wrapper.findComponent({ ref: "list" }).vm.$emit("select", options[0]);
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.showMenu).toBe(false);
    expect(document.activeElement).toBe(wrapper.find("input").element);
    expect(wrapper.emitted("update:modelValue")[0][0]).toStrictEqual(
      options[0]
    );
  });
  it("should not open menu when focused (tabbed to)", async () => {
    const wrapper = getFullWrapper({
      props: {
        options,
      },
    });
    await wrapper.find("input").trigger("focus");
    expect(wrapper.vm.showMenu).toBe(false);
  });
  it("should select item that starts with a character sequence when it is quickly typed", async () => {
    jest.useFakeTimers();

    const wrapper = getFullWrapper({
      props: {
        options: options.concat([{ _text: "ab" }]),
      },
    });
    await wrapper.find("input").trigger("focus");
    await wrapper.find("input").trigger("keydown.enter");
    await wrapper.find("input").trigger("keydown", {
      keyCode: 65,
    });
    await wrapper.find("input").trigger("keydown", {
      keyCode: 66,
    });
    jest.runAllTimers();
    expect(wrapper.vm.internalValue._text.indexOf("ab")).toBe(0);
  });
  describe("accessibility", () => {
    it("should be accessible in normal state", async () => {
      jest.useRealTimers();

      const wrapper = getFullWrapper();
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
    it("should be accessible in open state with value", async () => {
      jest.useRealTimers();

      const wrapper = getFullWrapper({
        props: {
          options,
          modelValue: options[0],
        },
      });
      await wrapper.find("input").trigger("focus");
      await wrapper.find("input").trigger("keydown.enter");
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
  });
});
