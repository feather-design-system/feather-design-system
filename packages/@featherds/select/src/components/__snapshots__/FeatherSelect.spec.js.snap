// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatherSelect.vue should not raise the label when its empty 1`] = `
<div
  class="feather-select-container"
>
  <div
    class="feather-menu feather-select-menu-container"
    data-v-6f68d0ea=""
  >
    
    <div
      class="feather-input-wrapper-container feather-select-wrapper"
      data-v-47a74f22=""
      menu-trigger=""
    >
      <fieldset
        aria-hidden="true"
        class="feather-input-border"
        data-v-47a74f22=""
      >
        <legend
          data-v-47a74f22=""
        >
          Users
        </legend>
      </fieldset>
      <label
        class="feather-input-label"
        data-ref-id="feather-form-element-label"
        data-v-47a74f22=""
        for="feather-select-input"
        style="left: 16px;"
      >
        Users
      </label>
      <div
        class="feather-input-wrapper"
        data-v-47a74f22=""
      >
        <div
          class="prefix"
          data-v-47a74f22=""
        >
          
          
        </div>
        
        <input
          aria-describedby=" feather-select-description"
          aria-disabled="false"
          aria-haspopup="listbox"
          aria-invalid="false"
          class="feather-select-input"
          data-ref-id="feather-input"
          id="feather-select-input"
          readonly=""
        />
        
        <div
          class="post"
          data-v-47a74f22=""
        >
          <!---->
          <!---->
          
          <svg
            aria-hidden="true"
            class="feather-select-icon feather-icon feather-select-icon"
            data-v-0aa06a12=""
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.7,9.29a1,1,0,0,0,0,1.41L12,16l5.3-5.3a1,1,0,0,0-1.41-1.41L12,13.17,8.11,9.29A1,1,0,0,0,6.7,9.29Z"
            />
          </svg>
          
        </div>
      </div>
    </div>
    
    <div
      class="feather-menu-dropdown bottom-left"
      data-v-6f68d0ea=""
      id="feather-menu-dropdown"
      style="display: none;"
    >
      
      <ul
        aria-activedescendant="feather-select-active"
        aria-label="Users"
        class="feather-list feather-select-options-list"
        data-ref-id="feather-select-list"
        data-v-4ef0e6d6=""
        role="listbox"
        tabindex="-1"
      >
        
        
      </ul>
      
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    data-v-764b8d6b=""
    id="feather-select-description"
    style="display: none;"
  >
    <!---->
    <!---->
    
    
  </div>
</div>
`;

exports[`FeatherSelect.vue should raise the label when it has a value 1`] = `
<div
  class="feather-select-container"
>
  <div
    class="feather-menu feather-select-menu-container"
    data-v-6f68d0ea=""
  >
    
    <div
      class="feather-input-wrapper-container raised feather-select-wrapper"
      data-v-47a74f22=""
      menu-trigger=""
    >
      <fieldset
        aria-hidden="true"
        class="feather-input-border"
        data-v-47a74f22=""
      >
        <legend
          data-v-47a74f22=""
        >
          Users
        </legend>
      </fieldset>
      <label
        class="feather-input-label"
        data-ref-id="feather-form-element-label"
        data-v-47a74f22=""
        for="feather-select-input"
        style="left: 12px;"
      >
        Users
      </label>
      <div
        class="feather-input-wrapper"
        data-v-47a74f22=""
      >
        <div
          class="prefix"
          data-v-47a74f22=""
        >
          
          
        </div>
        
        <input
          aria-describedby=" feather-select-description"
          aria-disabled="false"
          aria-haspopup="listbox"
          aria-invalid="false"
          class="feather-select-input"
          data-ref-id="feather-input"
          id="feather-select-input"
          readonly=""
        />
        
        <div
          class="post"
          data-v-47a74f22=""
        >
          <!---->
          <!---->
          
          <svg
            aria-hidden="true"
            class="feather-select-icon feather-icon feather-select-icon"
            data-v-0aa06a12=""
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.7,9.29a1,1,0,0,0,0,1.41L12,16l5.3-5.3a1,1,0,0,0-1.41-1.41L12,13.17,8.11,9.29A1,1,0,0,0,6.7,9.29Z"
            />
          </svg>
          
        </div>
      </div>
    </div>
    
    <div
      class="feather-menu-dropdown bottom-left"
      data-v-6f68d0ea=""
      id="feather-menu-dropdown"
      style="display: none;"
    >
      
      <ul
        aria-activedescendant="feather-select-active"
        aria-label="Users"
        class="feather-list feather-select-options-list"
        data-ref-id="feather-select-list"
        data-v-4ef0e6d6=""
        role="listbox"
        tabindex="-1"
      >
        
        
      </ul>
      
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    data-v-764b8d6b=""
    id="feather-select-description"
    style="display: none;"
  >
    <!---->
    <!---->
    
    
  </div>
</div>
`;

exports[`FeatherSelect.vue should render in disabled state 1`] = `
<div
  class="feather-select-container"
>
  <div
    class="feather-menu feather-select-menu-container"
    data-v-6f68d0ea=""
  >
    
    <div
      class="feather-input-wrapper-container disabled feather-select-wrapper"
      data-v-47a74f22=""
      menu-trigger=""
    >
      <fieldset
        aria-hidden="true"
        class="feather-input-border"
        data-v-47a74f22=""
      >
        <legend
          data-v-47a74f22=""
        >
          Users
        </legend>
      </fieldset>
      <label
        class="feather-input-label"
        data-ref-id="feather-form-element-label"
        data-v-47a74f22=""
        for="feather-select-input"
        style="left: 16px;"
      >
        Users
      </label>
      <div
        class="feather-input-wrapper"
        data-v-47a74f22=""
      >
        <div
          class="prefix"
          data-v-47a74f22=""
        >
          
          
        </div>
        
        <input
          aria-describedby=" feather-select-description"
          aria-disabled="true"
          aria-haspopup="listbox"
          aria-invalid="false"
          class="feather-select-input"
          data-ref-id="feather-input"
          disabled=""
          id="feather-select-input"
          readonly=""
        />
        
        <div
          class="post"
          data-v-47a74f22=""
        >
          <!---->
          <!---->
          
          <svg
            aria-hidden="true"
            class="feather-select-icon feather-icon feather-select-icon"
            data-v-0aa06a12=""
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.7,9.29a1,1,0,0,0,0,1.41L12,16l5.3-5.3a1,1,0,0,0-1.41-1.41L12,13.17,8.11,9.29A1,1,0,0,0,6.7,9.29Z"
            />
          </svg>
          
        </div>
      </div>
    </div>
    
    <div
      class="feather-menu-dropdown bottom-left"
      data-v-6f68d0ea=""
      id="feather-menu-dropdown"
      style="display: none;"
    >
      
      <ul
        aria-activedescendant="feather-select-active"
        aria-label="Users"
        class="feather-list feather-select-options-list"
        data-ref-id="feather-select-list"
        data-v-4ef0e6d6=""
        role="listbox"
        tabindex="-1"
      >
        
        
      </ul>
      
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    data-v-764b8d6b=""
    id="feather-select-description"
    style="display: none;"
  >
    <!---->
    <!---->
    
    
  </div>
</div>
`;

exports[`FeatherSelect.vue should render in error state 1`] = `
<div
  class="feather-select-container"
>
  <div
    class="feather-menu feather-select-menu-container"
    data-v-6f68d0ea=""
  >
    
    <div
      class="feather-input-wrapper-container error feather-select-wrapper"
      data-v-47a74f22=""
      menu-trigger=""
    >
      <fieldset
        aria-hidden="true"
        class="feather-input-border"
        data-v-47a74f22=""
      >
        <legend
          data-v-47a74f22=""
        >
          Users
        </legend>
      </fieldset>
      <label
        class="feather-input-label"
        data-ref-id="feather-form-element-label"
        data-v-47a74f22=""
        for="feather-select-input"
        style="left: 16px;"
      >
        Users
      </label>
      <div
        class="feather-input-wrapper"
        data-v-47a74f22=""
      >
        <div
          class="prefix"
          data-v-47a74f22=""
        >
          
          
        </div>
        
        <input
          aria-describedby=" feather-select-description"
          aria-disabled="false"
          aria-haspopup="listbox"
          aria-invalid="true"
          class="feather-select-input"
          data-ref-id="feather-input"
          id="feather-select-input"
          readonly=""
        />
        
        <div
          class="post"
          data-v-47a74f22=""
        >
          <!---->
          <svg
            aria-hidden="true"
            class="error-icon hide-when-disabled feather-icon error-icon hide-when-disabled"
            data-v-0aa06a12=""
            data-v-329cf64a=""
            data-v-47a74f22=""
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12,2A10,10,0,1,0,22,12,10,10,0,0,0,12,2Zm0,18a8,8,0,1,1,8-8A8,8,0,0,1,12,20Z"
            />
            <rect
              height="7"
              rx="1"
              width="2"
              x="11"
              y="7"
            />
            <rect
              height="2"
              rx="0.65"
              width="2"
              x="11"
              y="15"
            />
          </svg>
          
          <svg
            aria-hidden="true"
            class="feather-select-icon feather-icon feather-select-icon"
            data-v-0aa06a12=""
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.7,9.29a1,1,0,0,0,0,1.41L12,16l5.3-5.3a1,1,0,0,0-1.41-1.41L12,13.17,8.11,9.29A1,1,0,0,0,6.7,9.29Z"
            />
          </svg>
          
        </div>
      </div>
    </div>
    
    <div
      class="feather-menu-dropdown bottom-left"
      data-v-6f68d0ea=""
      id="feather-menu-dropdown"
      style="display: none;"
    >
      
      <ul
        aria-activedescendant="feather-select-active"
        aria-label="Users"
        class="feather-list feather-select-options-list"
        data-ref-id="feather-select-list"
        data-v-4ef0e6d6=""
        role="listbox"
        tabindex="-1"
      >
        
        
      </ul>
      
    </div>
  </div>
  <div
    class="feather-input-sub-text"
    data-v-764b8d6b=""
    id="feather-select-description"
  >
    <!---->
    <div
      aria-live="assertive"
      class="feather-input-error"
      data-ref-id="feather-form-element-error"
      data-v-764b8d6b=""
    >
      ERROR
    </div>
    
    
  </div>
</div>
`;
